Sources/FrontEnd/TypeChecking/TypeChecker.swift

```diff
@@ -3646,9 +3627,41 @@ public struct TypeChecker {
     return r
   }
 
+  /// Returns the declarations in `decls` whose names match `n`.
+  ///
+  /// - Requires: The base name of all declarations in `decls` is equal to `n-stem`
+  mutating func filter(decls: DeclSet, matching n: Name) -> [AnyDeclID] {
+    var matches = Array(decls)
+
+    // Filter out candidates whose argument labels do not match.
+    if !n.labels.isEmpty {
+      filter(decls: &matches, withLabels: n.labels)
+    }
+
+    // Filter out candidates whose operator notation does not match.
+    if let notation = n.notation {
+      filter(decls: &matches, withNotation: notation)
+    }
+
+    // If the looked up name has an introducer, select the corresponding implementation.
+    if let introducer = n.introducer {
+      matches = matches.compactMap({ (match) -> AnyDeclID? in
+        guard
+          let decl = program.ast[NodeID<MethodDecl>(match)],
+          let impl = decl.impls.first(where: { (i) in
+            program.ast[i].introducer.value == introducer
+          })
+        else { return nil }
+        return AnyDeclID(impl)
+      })
+    }
+
+    return matches
+  }
+
   /// Filters the function, method, and subscript declarations in `decls` whose argument labels
   /// match `labels`.
-  mutating func filter(decls: inout [AnyDeclID], withLabels labels: [String?]) {
+  private mutating func filter(decls: inout [AnyDeclID], withLabels labels: [String?]) {
```

Comment:
```suggestion
  private mutating func remove(elementsNotLabeled desiredLabels: [String?], from decls: inout [AnyDeclID]) {
```

Of course, I am obliged to ask if this should be a method of `[AnyDeclID]`.  Presumably you could use `remove(where:)` and have the typechecker supply a suitable predicate?

Reply:
The reason is that `self` (i.e. the `TypeChecker`) has to be projected mutably in the function and that this filter is only relevant in the context of the type checker. It feels like a lot of boilerplate to declare a conditional extension of `[AnyDecl]` as opposed to simply declaring the method here. I'm not opposed to it though, tradeoffs I guess.

Perhaps `remove(where:)` sidesteps the issue because `self` gets projected in the closure implicitly. The issue is that the predicate is a bit long for an inline closure, and declaring a method just to be a predicate of a single call to `remove(where:)` seems more convoluted than the current solution.

Let me know what you think.

Reply:
One of the problems with the name `filter` is that we talk about “filtering out” things that are removed, and yet the closure identifies things that will be kept.  That's what made me start on this, but not what I think is the most important issue: it's that we should use the existing algorithms where they are appropriate.  They are valuable abstractions, and that we use them is part of the Better Code philosophy.  It's very odd in that context to see something removing elements from a collection but not using the standard algorithm.  I think it's worth trying the `remove(where:)` formulation.